### PR TITLE
[WIP] feat: Self-serve Manage Form Ownership

### PR DESCRIPTION
**Pull request by dsamojlenko**
# Summary | Résumé

Self-serve Manage Form Ownership

| Access from Responses page  | Manage form access dialog    | Invite form |
| ----------------------------------------------- | -------------------------------------------- | ----- |
| image | image | image |

| Re-invite | Cancel invitation |
| ------------------------- | ----------------- |
|  Screen Recording 2024-09-26 at 4 21 49 PM | Screen Recording 2024-09-26 at 4 26 24 PM |

| Remove existing owner | Adding invitees + validation |
| ----- | ----- | 
| Screen Recording 2024-09-26 at 4 27 47 PM | Screen Recording 2024-09-26 at 4 35 24 PM |

| Accept or decline an invitation |
| ----- |
| Screen Recording 2024-09-26 at 4 57 28 PM |

## Test instructions

Using email aliases (ie, firstname.lastname+alias@cds-snc.ca) can be very helpful in testing this PR. If you can run the service locally, consider the Notify Interceptor.

Note that invitations are tied to a user's email address, and can be sent to someone whether they have an existing account or not.

When adding invitees, the following validations are run on email addresses:
- Must be from same domain as logged-in user
- Must be a government email address
- Must not already have access to the form

### Inviting a non-Forms user
- If an invitee does not have an account already, they will be sent an email referring them to the Registration form.
- After successfully registering they will land on their /forms page, where all invitations are displayed.
- If they accept the invitation, a notification goes to all existing owners to notify of the ownership change.

### Inviting a Forms user
- If an invitee already has an account, they will be sent an email instructing them to login to Forms to accept the invitation
- After successfully logging in, they will land on their /forms page, where all invitations are displayed.
- If they accept the invitation, a notification is sent to all existing owners to notify of the ownership change.

### Removing an owner
In the Manage Ownership dialog, existing owners are displayed. To remove an owner, press the Delete button and then Confirm that you want to delete the user. 

You will not be able to remove an Owner if:
- That owner is you
- There is only one owner remaining on the form

When an owner is removed, notifications are sent to the remaining owners, as well as the owner that was removed to confirm the ownership change.

### Cancelling an invitation
In the Manage Ownership dialog, previously sent invitations are displayed. To cancel an invitation, press the Cancel button and Confirm that you want to cancel. The Invitation is deleted, and no notifications are sent.

### Re-sending expired or lost invitations
In the Manage Ownership dialog, invitations that have been expired will be identified as "expired." To re-send any notification, press the resend button and the email address will be added to the Invitee list, then you can continue on the regular invitation process

### CODE REVIEW

**Comment by anikbrazeau** [comment]
Fixes #3127

File: i18n/translations/en/manage-form-access.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  "manageOwnership": "Manage ownership",
+  "mustBeAGovernmentAddress": "Must be a government email address",
+  "next": "Next",
+  "optionalMessage": "Optional message",
+  "peopleWithAccess": "People with access",
+  "remove": "Remove",
+  "resend": "Resend",
+  "templateNotFound": "Template not found {{templateId}}",
+  "userAlreadyHasAccess": "{{email}} already has access"
+}
```
Comment (anikbrazeau) [inline i18n/translations/en/manage-form-access.json:30]:
```suggestion
{
  "accessControlError": "You do not have permission to manage access to responses.",
  "addPeopleToShareAccess": "Share access",
  "areYouSure": "Are you sure?",
  "back": "Back",
  "cancel": "Cancel",
  "confirmError": "Something went wrong",
  "delete": "Delete",
  "deleteInvitation": "Delete invitation",
  "downloadForm": "Download form",
  "emailAlreadyInList": "{{email}} was already added",
  "error": "There was an error.",
  "invalidEmail": "{{email}} is not a valid email address",
  "invitationFailed": "Unable to send invitation to {{email}}",
  "invitationSent": "Invitation sent",
  "invite": "Invite",
  "invited": "Invited",
  "inviteToForm": "Invite to manage responses",
  "loading": "Loading...",
  "manageFormAccess": "Manage access",
  "manageOwnership": "Manage ownership",
  "mustBeAGovernmentAddress": "Add people with their government email address",
  "next": "Continue",
  "optionalMessage": "Message (optional)",
  "peopleWithAccess": "People with access",
  "remove": "Remove",
  "resend": "Re-send",
  "templateNotFound": "Template not found {{templateId}}",
  "userAlreadyHasAccess": "{{email}} already has access"
}
```

Comment (anikbrazeau) [reply i18n/translations/en/manage-form-access.json:30]: What does "Template not found" mean? Is that an internal error?

File: i18n/translations/fr/manage-form-access.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  "manageOwnership": "Manage ownership",
+  "mustBeAGovernmentAddress": "Must be a government email address",
+  "next": "Next",
+  "optionalMessage": "Optional message",
+  "peopleWithAccess": "People with access",
+  "remove": "Remove",
+  "resend": "Resend",
+  "templateNotFound": "Template not found {{templateId}}",
+  "userAlreadyHasAccess": "{{email}} already has access"
+}
```
Comment (anikbrazeau) [inline i18n/translations/fr/manage-form-access.json:30]:
```suggestion
{
 "accessControlError": "Vous n'avez pas la permission de gérer l'accès aux réponses.",
  "addPeopleToShareAccess": "Partager l'accès",
  "areYouSure": "Êtes-vous certain?",
  "back": "Retourner",
  "cancel": "Annuler",
  "confirmError": "Quelque chose a mal tourné",
  "delete": "Supprimer",
  "deleteInvitation": "Supprimer l'invitation",
  "downloadForm": "Télécharger le formulaire",
  "emailAlreadyInList": "{{email}} a déjà été ajouté",
  "error": "Il y a eu une erreur.",
  "invalidEmail": "{{email}} n'est pas une adresse courriel valide",
  "invitationFailed": "Impossible d'envoyer une invitation à {{email}}",
  "invitationSent": "Invitation envoyée",
  "invite": "Inviter",
  "invited": "Invité",
  "inviteToForm": "Inviter à gérer les réponses",
  "loading": "Téléchargement...",
  "manageFormAccess": "Gérer l'accès",
  "manageOwnership": "Gérer le propriétaire",
  "mustBeAGovernmentAddress": "Ajouter des personnes avec leur adresse courriel du gouvernement",
  "next": "Continuer",
  "optionalMessage": "Message (optionnel)",
  "peopleWithAccess": "Personnes ayant accès",
  "remove": "Supprimer",
  "resend": "Renvoyer",
  "templateNotFound": "Gabarit non trouvé {{templateId}}",
  "userAlreadyHasAccess": "{{email}} a déjà accès"
}
```

File: i18n/translations/en/manage-form-access.json (hunk 1 above)
Comment (dsamojlenko) [reply i18n/translations/en/manage-form-access.json]: It's a potential but unlikely error. Basically before most actions are taken, we do two basic checks - that the user requesting the invite exists, and that the template exists.

File: app/(gcforms)/[locale]/(form administration)/forms/components/Invitations/index.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    fetchInvitations();
+  };
+
+  return (
+    <div className="mb-4 flex flex-col gap-2">
+      {invitations.map((invitation: Invitation) => {
+        return (
+          <div key={invitation.id} className="rounded-md bg-violet-50 p-4">
+            <StarIcon className="mr-1 inline-block size-8" />
+            You have been invited to access a form -{" "}
```
Comment (timarney) [inline app/(gcforms)/[locale]/(form administration)/forms/components/Invitations/index.tsx:37]: Needs translations ?

File: lib/types/user-types.ts [hunk 4]
```diff
@@ -3,7 +3,7 @@ import { User } from "@prisma/client";
 export interface FormOwner {
   id: string;
   email: string;
-  name?: string;
+  name?: string | null;
```
Comment (timarney) [inline lib/types/user-types.ts:6]: What's the case for null here?

File: app/(gcforms)/[locale]/(form administration)/forms/components/Invitations/index.tsx (hunk 3 above)
Comment (dsamojlenko) [reply app/(gcforms)/[locale]/(form administration)/forms/components/Invitations/index.tsx:37]: yup

File: i18n/translations/fr/manage-form-access.json [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  "delete": "Supprimer",
+  "deleteInvitation": "Supprimer l'invitation",
+  "downloadForm": "Télécharger le formulaire",
+  "emailAlreadyInList": "{{email}} a déjà été ajouté",
+  "error": "Il y a eu une erreur.",
+  "expired": "Expiré",
+  "invalidEmail": "{{email}} n'est pas une adresse courriel valide",
+  "invitationFailed": "Impossible d'envoyer une invitation à {{email}}",
+  "invitationSent": "Invitation envoyée",
+  "invitationToAccessForm": "You have been invited to access a form - <accept>accept</accept> or <decline>decline</decline> invitation.",
```
Comment (anikbrazeau) [inline i18n/translations/fr/manage-form-access.json:17]:
```suggestion
  "invitationToAccessForm": "Vous avez été invité à accéder aux réponses de formulaire - Veuillez <accept>accepter</accept> ou <decline>refuser</decline> l'invitation.",
```

File: i18n/translations/en/manage-form-access.json [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  "addPeopleToShareAccess": "Share access",
+  "areYouSure": "Are you sure?",
+  "back": "Back",
+  "cancel": "Cancel",
+  "confirmError": "Something went wrong",
+  "delete": "Delete",
+  "deleteInvitation": "Delete invitation",
+  "downloadForm": "Download form",
+  "emailAlreadyInList": "{{email}} was already added",
+  "emailDomainMismatch": "{{email}} Email domain mismatch",
```
Comment (anikbrazeau) [inline i18n/translations/en/manage-form-access.json:12]:
```suggestion
  "emailDomainMismatch": "{{email}} does not appear to be from your organization. You can only add collaborators with the same email domain.",
```

File: i18n/translations/fr/manage-form-access.json [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  "addPeopleToShareAccess": "Partager l'accès",
+  "areYouSure": "Êtes-vous certain?",
+  "back": "Retourner",
+  "cancel": "Annuler",
+  "confirmError": "Quelque chose a mal tourné",
+  "delete": "Supprimer",
+  "deleteInvitation": "Supprimer l'invitation",
+  "downloadForm": "Télécharger le formulaire",
+  "emailAlreadyInList": "{{email}} a déjà été ajouté",
+  "emailDomainMismatch": "{{email}} Email domain mismatch",
```
Comment (anikbrazeau) [inline i18n/translations/fr/manage-form-access.json:12]:
```suggestion
  "emailDomainMismatch": "{{email}} ne semble pas appartenir à votre organisation. Vous ne pouvez ajouter des collaborateurs qu'avec le même domaine d'adresse courriel.",
```

File: lib/invitations/emailTemplates/inviteToCollaborate.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+[Open ${formTitleEn || formTitleFr}](${formUrlEn})
+
+===
+
+Vous avez été invité à accéder aux réponses de formulaire dans Formulaires GC par ${senderName}.
+
+${message}
+
+[Ouvrir ${formTitleFr || formTitleEn}](${formUrlFr})
+`;
```
Comment (anikbrazeau) [inline lib/invitations/emailTemplates/inviteToCollaborate.ts:35]:
```suggestion
(la version française suit)

You’ve been invited to access form responses in GC Forms by ${senderName}.

${message}

[Open ${formTitleEn}](${formUrlEn})

===

Vous avez été invité à accéder aux réponses de formulaire dans Formulaires GC par ${senderName}.

${message}

[Ouvrir ${formTitleFr}](${formUrlFr})
`;
```

File: lib/invitations/emailTemplates/ownerAddedNotification.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+): string => `
+(la version française suit)
+
+${formOwner} has been added to the form ${formTitleEn || formTitleFr}:
+
+======
+
+${formOwner} a été ajouté au formulaire ${formTitleFr || formTitleEn}:
+
+`;
```
Comment (anikbrazeau) [inline lib/invitations/emailTemplates/ownerAddedNotification.ts:22]:
```suggestion
(la version française suit)

${formOwner} has been added to the form: ${formTitleEn}:

======

${formOwner} a été ajouté au formulaire : ${formTitleFr}:

`;
```

File: lib/invitations/emailTemplates/ownerRemovedNotification.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  formOwner: string
+): string => `
+(la version française suit)
+
+${formOwner} no longer has access to responses for the form: ${formTitleEn || formTitleFr}.
+
+======
+
+${formOwner} n’a plus accès aux réponses du formulaire ${formTitleFr || formTitleEn}.:
+`;
```
Comment (anikbrazeau) [inline lib/invitations/emailTemplates/ownerRemovedNotification.ts:21]:
```suggestion
(la version française suit)

${formOwner} no longer has access to responses for the form: ${formTitleFr}.

======

${formOwner} n’a plus accès aux réponses du formulaire : ${formTitleFr}.:
`;
```

File: lib/invitations/emailTemplates/youHaveBeenRemovedNotification.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+): string => `
+(la version française suit)
+
+Your access was removed for the form: ${formTitleEn || formTitleFr}.
+
+======
+
+Vous n’avez plus accès au formulaire : ${formTitleFr || formTitleEn}
+
+`;
```
Comment (anikbrazeau) [inline lib/invitations/emailTemplates/youHaveBeenRemovedNotification.ts:20]:
```suggestion
(la version française suit)

Your access was removed for the form: ${formTitleEn}.

======

Vous n’avez plus accès au formulaire : ${formTitleFr}

`;
```

File: lib/types/user-types.ts (hunk 4 above)
Comment (dsamojlenko) [reply lib/types/user-types.ts:6]: in the new invitations lib PR, I've modified the code so we don't need this anymore.

**Comment by dsamojlenko** [comment]
superceded by: https://github.com/cds-snc/platform-forms-client/pull/4386 and https://github.com/cds-snc/platform-forms-client/pull/4392

**Comment by dsamojlenko** [comment]
superceded by: https://github.com/cds-snc/platform-forms-client/pull/4386 and https://github.com/cds-snc/platform-forms-client/pull/4392

**Pull request closed without merging**